test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorPageTest.cs

```diff
@@ -225,6 +225,47 @@ public static TheoryData<HtmlEncoder> HtmlEncoderData
             await page.ExecuteAsync();
         }
 
+        [Fact]
+        public async Task EndWriteTagHelperAttribute_RestoresPageWriter()
+        {
+            // Arrange
+            var page = CreatePage(v =>
+            {
+                v.BeginWriteTagHelperAttribute();
+                v.Write("Hello World!");
+                var returnValue = v.EndWriteTagHelperAttribute();
+            });
+            var originalWriter = page.Output;
+
+            // Act
+            await page.ExecuteAsync();
+
+            // Assert
+            Assert.NotNull(originalWriter);
+            Assert.Same(originalWriter, page.Output);
+        }
+
+        [Fact]
+        public async Task EndWriteTagHelperAttribute_ReturnsAppropriateContent()
+        {
+            // Arrange
+            var viewContext = CreateViewContext();
+            var page = CreatePage(v =>
+            {
+                v.HtmlEncoder = new HtmlTestEncoder();
+                v.BeginWriteTagHelperAttribute();
+                v.Write("Hello World!");
+                var returnValue = v.EndWriteTagHelperAttribute();
+
+                // Assert
+                var content = Assert.IsType<string>(returnValue);
+                Assert.Equal("HtmlEncode[[Hello World!]]", content);
```

Comment:
Assertions in lambdas may lead to unintelligible test failures. Suggest checking quickly e.g. mess up the expected string, run test, and confirm the failure is useful.